README.md
> *NOTE: This is a fork of [Clément Guillemain](https://github.com/Sybio)'s nice [GifCreator class](https://github.com/Sybio/GifCreator), with some minor bug fixes and most pull requests merged together.

# ================================
# GifCreator
# ================================

GifCreator is a PHP class to create animated GIF from multiple images

### For what ?

This class helps you to create an animated GIF image: give multiple images and their duration and that's it !

### Usage

**1 - Creation:**

```php
// Create an array containing file paths, resource var (initialized with imagecreatefromXXX),
// image URLs or even binary code from image files.
// All sorted in order to appear.
$frames = array(
    imagecreatefrompng("/../images/pic1.png"), // Resource var
    "/../images/pic2.png", // Image file path
    file_get_contents("/../images/pic3.jpg"), // Binary source code
    'http://thisisafakedomain.com/images/pic4.jpg', // URL
);

// Create an array containing the duration (in centiseconds (hundredths of a second)) of each frames (in order too)
$durations = array(40, 80, 40, 20);

// Initialize and create the GIF !
$gc = new GifCreator\GifCreator();
$gc->create($frames, $durations, 5);
```
The 3rd parameter of create() method allows you to choose the number of loop of your animated gif before it stops.
In the previous example, I chose 5 loops. Set 0 (zero) to get an infinite loop.

**2 - Get the result:**

You can now get the animated GIF binary:

```php
$gifBinary = $gc->getGif();
```

Then you can show it in the navigator:

```php
header('Content-type: image/gif');
header('Content-Disposition: filename="butterfly.gif"');
echo $gifBinary;
exit;
```

Or save it in a folder as a GIF:

```php
file_put_contents('/myfolder/animated_picture.gif', $gifBinary);
```

### Behavior

- The transparency is based on the first given frame. It will be saved only if you give multiple frames with same transparent background.
- The dimensions of the generated GIF are based on the first frame. If you need to resize your frames to get the same dimension, you can use
this class: https://github.com/Sybio/ImageWorkshop

### About

The class reuses some part of code of "GIFEncoder.class.php" by László Zsidi (thanks to him).


CHANGES.txt
TODO:

! Allow image dirs within the input list (array), too, not only instead of it.

! In local testing I had a strange issue with by-frame delays: the first image eats
  up the first two delays, and is shown for the sum of those two numbers... :-o
  -> And I seem to be unable to reproduce it now! :-o

* Warn about explicit sorting if per-frame delays are specified for a dir source?

* Sync the index.php (demo & quickstart frontend) with the examples in README.md 
  AND ALSO THE ONE AT SO!

* To avoid unwanted "in-band" PHP output, I should probably disable display_errors()
  in the whole module (restoring it on normal exit)...


------------------------------------------------------------------------------------
DONE:

-> Release notes: https://github.com/lunakid/AnimGif/releases/
-> See also the GitHub commit history: https://github.com/lunakid/AnimGif/commits/


test/index.php
<?php

	namespace GifCreator;
	
	require_once '../src/GifCreator/GifCreator.php';

	// Create an array containing file paths, resource var (initialized with imagecreatefromXXX), 
	// image URLs or even binary code from image files.
	// All sorted in order to appear.
	$frames = array(
	    "1.png",
	    "2.png",
	);

	// Create an array containing the duration (in millisecond) of each frames (in order too)
	$durations = array(40, 80, 40);

	// Initialize and create the GIF !
	$gc = new GifCreator();
	$gc->create($frames, $durations, 5);


	$gifBinary = $gc->getGif();

	if (isset($_GET['save'])) {
		file_put_contents('final.gif', $gifBinary);
	}else{
		header('Content-type: image/gif');
		header('Content-Disposition: filename="butterfly.gif"');
		echo $gifBinary;
		exit;
	}



src/GifCreator/GifCreator.php
<?php

namespace GifCreator;

/**
 * Create an animated GIF from multiple images
 *
 * @version 1.0
 * @link https://github.com/Sybio/GifCreator
 * @author Sybio (Clément Guillemain  / @Sybio01)
 * @license http://opensource.org/licenses/gpl-license.php GNU Public License
 * @copyright Clément Guillemain
 */
class GifCreator
{
    /**
     * @var string The gif string source (old: this->GIF)
     */
    private $gif;

    /**
     * @var string Encoder version (old: this->VER)
     */
    private $version;

    /**
     * @var boolean Check the image is build or not (old: this->IMG)
     */
    private $imgBuilt;

    /**
     * @var array Frames string sources (old: this->BUF)
     */
    private $frameSources;

    /**
     * @var integer Gif loop (old: this->LOP)
     */
    private $loop;

    /**
     * @var integer Gif dis (old: this->DIS)
     */
    private $dis;

    /**
     * @var integer Gif color (old: this->COL)
     */
    private $colour;

    /**
     * @var array (old: this->ERR)
     */
    private $errors;

    // Methods
    // ===================================================================================

    /**
     * Constructor
     */
    public function __construct()
    {
        $this->reset();

        // Static data
        $this->version = 'GifCreator: Under development';
        $this->errors = array(
            'ERR00' => 'Does not support single image/duration input.',
            'ERR01' => 'Source is not a GIF image.',
            'ERR02' => 'You have to give resource image variables, image URL or image binary sources in $frames array.',
            'ERR03' => 'Does not make animation from animated GIF source.',
        );
    }

    /**
     * Create the GIF string (old: GIFEncoder)
     *
     * @param array $frames An array of frame: can be file paths, resource image variables, binary sources or image URLs
     * @param array $durations An array containing the duration of each frame (in centiseconds)
     * @param integer $loop Number of GIF loops before stopping animation (Set 0 to get an infinite loop)
     * @return string The GIF string source
     * @throws \Exception
     */
    public function create($frames = array(), $durations = array(), $loop = 0)
    {
        if (!is_array($frames) && !is_array($durations)) {

            throw new \Exception($this->version . ': ' . $this->errors['ERR00']);
        }

        $this->loop = ($loop > -1) ? $loop : 0;
        $this->dis = 2;

        for ($i = 0; $i < count($frames); $i++) {

            if (is_resource($frames[$i])) { // Resource var

                $resourceImg = $frames[$i];

                ob_start();
                imagegif($frames[$i]);
                $this->frameSources[] = ob_get_contents();
                ob_end_clean();

            } elseif (is_string($frames[$i])) { // File path or URL or Binary source code

                if ((substr($frames[$i], 0, 4) != 'GIF8') && file_exists($frames[$i]) || filter_var($frames[$i], FILTER_VALIDATE_URL)) { // File path

                    $frames[$i] = file_get_contents($frames[$i]);
                }

                $resourceImg = imagecreatefromstring($frames[$i]);

                ob_start();
                imagegif($resourceImg);
                $this->frameSources[] = ob_get_contents();
                ob_end_clean();

            } else { // Fail

                throw new \Exception($this->version . ': ' . $this->errors['ERR02']);//.' ('.$mode.')'?
            }

            if ($i == 0) {

                $colour = imagecolortransparent($resourceImg);
            }

            if (substr($this->frameSources[$i], 0, 6) != 'GIF87a' && substr($this->frameSources[$i], 0, 6) != 'GIF89a') {

                throw new \Exception($this->version . ': ' . $i . ' ' . $this->errors['ERR01']);
            }

            for ($j = (13 + 3 * (2 << (ord($this->frameSources[$i]{10}) & 0x07))), $k = true; $k; $j++) {

                switch ($this->frameSources[$i]{$j}) {

                    case '!':

                        if ((substr($this->frameSources[$i], ($j + 3), 8)) == 'NETSCAPE') {

                            throw new \Exception($this->version . ': ' . $this->errors['ERR03'] . ' (' . ($i + 1) . ' source).');
                        }

                        break;

                    case ';':

                        $k = false;
                        break;
                }
            }

            unset($resourceImg);
        }

        if (isset($colour)) {

            $this->colour = $colour;

        } else {

            $red = $green = $blue = 0;
            $this->colour = ($red > -1 && $green > -1 && $blue > -1) ? ($red | ($green << 8) | ($blue << 16)) : -1;
        }

        $this->gifAddHeader();

        for ($i = 0; $i < count($this->frameSources); $i++) {

            $this->addGifFrames($i, $durations[$i]);
        }

        $this->gifAddFooter();

        return $this->gif;
    }

    // Internals
    // ===================================================================================

    /**
     * Add the header gif string in its source (old: GIFAddHeader)
     */
    public function gifAddHeader()
    {
        if (ord($this->frameSources[0]{10}) & 0x80) {

            $cmap = 3 * (2 << (ord($this->frameSources[0]{10}) & 0x07));

            $this->gif .= substr($this->frameSources[0], 6, 7);
            $this->gif .= substr($this->frameSources[0], 13, $cmap);
            if ($this->loop != 1) {
                // Browsers interpret loop count differently, but for loop == 1
                // we can work around this by excluding loop count completely
                // since all browsers seem to handle "no loop count" as
                // "play once only".
                $this->gif .= "!\377\13NETSCAPE2.0\3\1" . $this->encodeAsciiToChar($this->loop) . "\0";
            }
        }
    }

    /**
     * Add the frame sources to the GIF string (old: GIFAddFrames)
     *
     * @param integer $i
     * @param integer $d
     */
    public function addGifFrames($i, $d)
    {
        $Locals_str = 13 + 3 * (2 << (ord($this->frameSources[$i]{10}) & 0x07));

        $Locals_end = strlen($this->frameSources[$i]) - $Locals_str - 1;
        $Locals_tmp = substr($this->frameSources[$i], $Locals_str, $Locals_end);

        $Global_len = 2 << (ord($this->frameSources[0]{10}) & 0x07);
        $Locals_len = 2 << (ord($this->frameSources[$i]{10}) & 0x07);

        $Global_rgb = substr($this->frameSources[0], 13, 3 * (2 << (ord($this->frameSources[0]{10}) & 0x07)));
        $Locals_rgb = substr($this->frameSources[$i], 13, 3 * (2 << (ord($this->frameSources[$i]{10}) & 0x07)));

        $Locals_ext = "!\xF9\x04" . chr(($this->dis << 2) + 0) . chr(($d >> 0) & 0xFF) . chr(($d >> 8) & 0xFF) . "\x0\x0";

        if ($this->colour > -1 && ord($this->frameSources[$i]{10}) & 0x80) {

            for ($j = 0; $j < (2 << (ord($this->frameSources[$i]{10}) & 0x07)); $j++) {

                if (ord($Locals_rgb{3 * $j + 0}) == (($this->colour >> 16) & 0xFF) &&
                    ord($Locals_rgb{3 * $j + 1}) == (($this->colour >> 8) & 0xFF) &&
                    ord($Locals_rgb{3 * $j + 2}) == (($this->colour >> 0) & 0xFF)
                ) {
                    $Locals_ext = "!\xF9\x04" . chr(($this->dis << 2) + 1) . chr(($d >> 0) & 0xFF) . chr(($d >> 8) & 0xFF) . chr($j) . "\x0";
                    break;
                }
            }
        }
        $Locals_img = '';

        switch ($Locals_tmp{0}) {

            case '!':

                $Locals_img = substr($Locals_tmp, 8, 10);
                $Locals_tmp = substr($Locals_tmp, 18, strlen($Locals_tmp) - 18);

                break;

            case ',':

                $Locals_img = substr($Locals_tmp, 0, 10);
                $Locals_tmp = substr($Locals_tmp, 10, strlen($Locals_tmp) - 10);

                break;
        }
        if (ord($this->frameSources[$i]{10}) & 0x80 && $this->imgBuilt) {

            if ($Global_len == $Locals_len) {

                if ($this->gifBlockCompare($Global_rgb, $Locals_rgb, $Global_len)) {

                    $this->gif .= $Locals_ext . $Locals_img . $Locals_tmp;

                } else {

                    $byte = ord($Locals_img{9});
                    $byte |= 0x80;
                    $byte &= 0xF8;
                    $byte |= (ord($this->frameSources[0]{10}) & 0x07);
                    $Locals_img{9} = chr($byte);
                    $this->gif .= $Locals_ext . $Locals_img . $Locals_rgb . $Locals_tmp;
                }

            } else {

                $byte = ord($Locals_img{9});
                $byte |= 0x80;
                $byte &= 0xF8;
                $byte |= (ord($this->frameSources[$i]{10}) & 0x07);
                $Locals_img{9} = chr($byte);
                $this->gif .= $Locals_ext . $Locals_img . $Locals_rgb . $Locals_tmp;
            }

        } else {

            $this->gif .= $Locals_ext . $Locals_img . $Locals_tmp;
        }

        $this->imgBuilt = true;
    }

    /**
     * Add the gif string footer char (old: GIFAddFooter)
     */
    public function gifAddFooter()
    {
        $this->gif .= ';';
    }

    /**
     * Compare two block and return the version (old: GIFBlockCompare)
     *
     * @param string $globalBlock
     * @param string $localBlock
     * @param integer $length
     *
     * @return integer
     */
    public function gifBlockCompare($globalBlock, $localBlock, $length)
    {
        for ($i = 0; $i < $length; $i++) {

            if ($globalBlock{3 * $i + 0} != $localBlock{3 * $i + 0} ||
                $globalBlock{3 * $i + 1} != $localBlock{3 * $i + 1} ||
                $globalBlock{3 * $i + 2} != $localBlock{3 * $i + 2}
            ) {

                return 0;
            }
        }

        return 1;
    }

    /**
     * Encode an ASCII char into a string char (old: GIFWord)
     *
     * $param integer $char ASCII char
     *
     * @param $char
     * @return string
     */
    public function encodeAsciiToChar($char)
    {
        return (chr($char & 0xFF) . chr(($char >> 8) & 0xFF));
    }

    /**
     * Reset and clean the current object
     */
    public function reset()
    {
        $this->frameSources = [];
        $this->gif = 'GIF89a'; // the GIF header
        $this->imgBuilt = false;
        $this->loop = 0;
        $this->dis = 2;
        $this->colour = -1;
    }

    // Getter / Setter
    // ===================================================================================

    /**
     * Get the final GIF image string (old: GetAnimation)
     *
     * @return string
     */
    public function getGif()
    {
        return $this->gif;
    }
}


src/GifCreator/AnimGif.php
<?php

/* CHANGES of the "lunakid fork":

See CHANGES.txt for TODO items and release notes!

DONE:
+ See the GitHub commit log for the most recent (& most detailed) change history!
--------
Old (pre-1.2) manual change log (partly parallel with the GitHub commit log):
+ Renamed encodeAsciiToChar() to word2bin() & fixed its description.
+ Added error + check: 'ERR04' => 'Loading from URLs is disabled by PHP.'.
+ file_exists() -> @is_readable() (Better to take no risk of any PHP output
  in a raw GIF HTTP response...)
+ Oops, also need to fix the default delay. And then also change it to 100ms.
  (Because browsers seem to have difficulties handling too fast animations.)
+ Anim delay doesn't seem to be set in ms, at all. :-o
  -> Yeah, they ARE NOT in millisecs! See: http://www.w3.org/Graphics/GIF/spec-gif89a.txt
  Fixing the docs.
1.1:
+ Fixed ERR01 "Source is not a GIF image.": there's a .png in the examples!
  -> And it does support non-GIF files actually!
  Moved the error check to resource inputs only, and changed it to
  "Resource is not a GIF image.".
+ create() should iterate $frames with foreach() not with for, assuming direct
  indexes from 0 to < count.
  (The array keys can be anything, and should not affect the results.)
+ Removed unused $mode from reporting ERR02.
+ $duration now has a meaningful default in create().
+ $frames was incorrectly made an optional arg. of create().
+ Support uniform timing without an array.
+ Separate ERR00 from $durations not being an array.
+ Fixed leftover $GIF_tim in create().
+ Renamed method getGif() to get().
+ Renamed class to AnimGif (from GifCreator).
+ Made $this->version a define (VERSION).
+ Made $this->errors really static (self::$errors).
+ Moved encodeAsciiToChar() out from the class to the namespace root as a general utility fn.
+ Moved getGif to the "public" section (and removed the Getter/Setter section).
+ Moved reset() closer up to the ctor.
+ Changed comments here & there.
+ Whitespaces: fixed some tab/space mismatch etc.
+ Changed {$i} indexes to [$i] in gifBlockCompare(), just for fun. ;)
*/

/**
 * Create an animated GIF from multiple images
 * 
 * @link https://github.com/lunakid/AnimGif
 * @author Sybio (Clément Guillemain / @Sybio01), lunakid (@GitHub, @Gmail, @SO etc.)
 * @license http://opensource.org/licenses/gpl-license.php GNU Public License
 * @copyright Clément Guillemain, Szabolcs Szász
 */

namespace GifCreator;

const VERSION = '1.3+';

class AnimGif
{
	const DEFAULT_DURATION = 10;

	/**
	* @var string: The generated (binary) image
	*/
	private $gif;

	/**
	* @var boolean: Has an image (frame) been added already?
	*/
	private $imgBuilt;

	/**
	* @var array or string: Frame sources like filenames, URLs, bin. data, or a folder name
	*/
	private $frameSources;

	/**
	* @var integer: Gif loop count
	*/
	private $loop;

	/**
	* @var integer: Gif frame disposal method
	*/
	private $dis;

	/**
	* @var integer: Gif transparent color index
	*/
	private $transparent_color;

	/**
	* @var array
	*/
	private static $errors;
 
	// Methods
	// ===================================================================================
    
	public function __construct()
	{
		$this->reset();

		self::$errors = array(
			'ERR00' => 'Need at least 2 frames for an animation.',
			'ERR01' => 'Resource is not a GIF image.',
			'ERR02' => 'Only image resource variables, file paths, URLs or binary bitmap data are accepted.',
			'ERR03' => 'Cannot make animation from animated GIF.',
			'ERR04' => 'Loading from URLs is disabled by PHP.',
			'ERR05' => 'Failed to load or invalid image (dir): "%s".',
		);
	}

	/**
	 * Create animated GIF from source images
	 * 
	 * @param array $frames The source iamges: can be a local dir path, or an array  
	 *                      of file paths, resource image variables, binary data or image URLs.
	 * @param array|number $durations The duration (in 1/100s) of the individual frames, 
	 *                      or a single integer for each one.
	 * @param integer $loop Number of loops before stopping the animation (set to 0 for infinite looping).
	 * 
	 * @return string The resulting GIF binary data.
	 */
	public function create($frames, $durations = self::DEFAULT_DURATION, $loop = 0)
	{
		$last_duration = self::DEFAULT_DURATION; // used only if $durations is an array
		
		$this->loop = ($loop > -1) ? $loop : 0; // Negatives would be fine for the GIF itself (all unsigned!), 
		                                        // but better not take a chance with the PHP bit-ops...
		$this->dis = 2; // "reset to bgnd." (http://www.matthewflickinger.com/lab/whatsinagif/animation_and_transparency.asp)

		// Check if $frames is a dir; get all files in ascending order if yes (else die):
		if (!is_array($frames)) {
			$frames_dir = $frames;
			if (@is_dir($frames_dir)) {
				if ($frames = scandir($frames_dir)) {
					$frames = array_filter($frames, function($x) { 
						// Should these two below be selectable?
						return $x[0] != "."; // Or: $x != "." && $x != "..";
					});

					array_walk($frames, function(&$x, $i) use ($frames_dir) { 
						$x = "$frames_dir/$x"; });
				}
			}
				
			if (!is_array($frames)) { // i.e. scandir() failed above!
				throw new \Exception(VERSION.': '
					. sprintf(self::$errors['ERR05'], $frames_dir)); // $frames is expected to be a string here; see the other ERR05 case!
			}
		}

		assert(is_array($frames));

		if (sizeof($frames) < 2) {
			throw new \Exception(VERSION.': '.self::$errors['ERR00']);
		}

		$i = 0;
		foreach ($frames as $frame) {
			if (is_resource($frame)) { // in-memory image resource (hopefully)

				$resourceImg = $frame;

				ob_start();
				imagegif($frame);
				$this->frameSources[] = ob_get_contents();
				ob_end_clean();

				if (substr($this->frameSources[$i], 0, 6) != 'GIF87a' && substr($this->frameSources[$i], 0, 6) != 'GIF89a') {
					throw new \Exception(VERSION.': '.$i.' '.self::$errors['ERR01']);
				}
	
			} elseif (is_string($frame)) { // file path, URL or binary data
			     
				if (@is_readable($frame)) { // file path
					$bin = file_get_contents($frame);                    
				} else if (filter_var($frame, FILTER_VALIDATE_URL)) {
 					if (ini_get('allow_url_fopen')) {
						$bin = @file_get_contents($frame);
					} else {
						throw new \Exception(VERSION.': '.$i.' '.self::$errors['ERR04']);
					}
				} else {
					$bin = $frame;
				}
				
				if (! ($bin && ($resourceImg = imagecreatefromstring($bin))) )
				{
					throw new \Exception(VERSION.': '.$i.' ' 
						. sprintf(self::$errors['ERR05'], substr($frame, 0, 200))); //!! $frame may be binary data, not a name!
				}

				ob_start();
				imagegif($resourceImg);
				$this->frameSources[] = ob_get_contents();
				ob_end_clean();
		 
			} else { // Fail
				throw new \Exception(VERSION.': '.self::$errors['ERR02']);
			}

			if ($i == 0) {
				$this->transparent_color = imagecolortransparent($resourceImg);
			}

			for ($j = (13 + 3 * (2 << (ord($this->frameSources[$i] { 10 }) & 0x07))), $k = TRUE; $k; $j++) {
			 
				switch ($this->frameSources[$i] { $j }) {
				    
					case '!':
		    				if ((substr($this->frameSources[$i], ($j + 3), 8)) == 'NETSCAPE') {
			    
							throw new \Exception(VERSION.': '.self::$errors['ERR03'].' ('.($i + 1).' source).');
						}
			
						break;
			
					case ';':
		    				$k = false;
						break;
				}
			}

			unset($resourceImg);

			++$i;
		}//foreach


		$this->gifAddHeader();

		for ($i = 0; $i < count($this->frameSources); $i++) {

			// Reuse the last delay if none has been specified for the current frame.
			if (is_array($durations)) {
				$d = (empty($durations[$i]) ? $last_duration : $durations[$i]);
				$last_duration = $d;
			} else {
				$d = $durations;
			}

			$this->addGifFrames($i, $d);
		}

		$this->gifAddFooter();

		return $this;
	}

	/**
	 * Get the resulting GIF image binary
	 * 
	 * @return string
	 */
	public function get()
	{
		return $this->gif;
	}
    
	/**
	 * Save the resulting GIF to a file.
	 * 
	 * @param $filename String Target file path
	 * 
	 * @return that of file_put_contents($filename)
	 */
	public function save($filename)
	{
		return file_put_contents($filename, $this->gif);
	}
    
	/**
	 * Clean-up the current object (also used by the ctor.)
	 */
	public function reset()
	{
		$this->frameSources = null;
		$this->gif = 'GIF89a'; // the GIF header
		$this->imgBuilt = false;
		$this->loop = 0;
		$this->dis = 2;
		$this->transparent_color = -1;
	}
	    
	// Internals
	// ===================================================================================

	/**
	 * Assemble the GIF header
	 */
	protected function gifAddHeader()
	{
		$cmap = 0;

		if (ord($this->frameSources[0] { 10 }) & 0x80) {
		  
			$cmap = 3 * (2 << (ord($this->frameSources[0] { 10 }) & 0x07));

			$this->gif .= substr($this->frameSources[0], 6, 7);
			$this->gif .= substr($this->frameSources[0], 13, $cmap);
			if ($this->loop !== 1) // Only add the looping extension if really looping
				$this->gif .= "!\xFF\x0BNETSCAPE2.0\x03\x01".word2bin($this->loop==0?0:$this->loop-1)."\x0";
		}
	}
    
	/**
	 * Add frame to the GIF data
	 * 
	 * @param integer $i: index of frame source
	 * @param integer $d: delay time (frame display duration)
	 */
	protected function addGifFrames($i, $d)
	{
		$Locals_str = 13 + 3 * (2 << (ord($this->frameSources[ $i ] { 10 }) & 0x07));

		$Locals_end = strlen($this->frameSources[$i]) - $Locals_str - 1;
		$Locals_tmp = substr($this->frameSources[$i], $Locals_str, $Locals_end);

		$Global_len = 2 << (ord($this->frameSources[0 ] { 10 }) & 0x07);
		$Locals_len = 2 << (ord($this->frameSources[$i] { 10 }) & 0x07);

		$Global_rgb = substr($this->frameSources[ 0], 13, 3 * (2 << (ord($this->frameSources[ 0] { 10 }) & 0x07)));
		$Locals_rgb = substr($this->frameSources[$i], 13, 3 * (2 << (ord($this->frameSources[$i] { 10 }) & 0x07)));

		$Locals_ext = "!\xF9\x04" . chr(($this->dis << 2) + 0) . word2bin($d) . "\x0\x0";

		if ($this->transparent_color > -1 && ord($this->frameSources[$i] { 10 }) & 0x80) {
		  
			for ($j = 0; $j < (2 << (ord($this->frameSources[$i] { 10 } ) & 0x07)); $j++) {
			 
				if (ord($Locals_rgb { 3 * $j + 0 }) == (($this->transparent_color >> 16) & 0xFF) &&
					ord($Locals_rgb { 3 * $j + 1 }) == (($this->transparent_color >> 8) & 0xFF) &&
					ord($Locals_rgb { 3 * $j + 2 }) == (($this->transparent_color >> 0) & 0xFF)
				) {
					$Locals_ext = "!\xF9\x04".chr(($this->dis << 2) + 1).chr(($d >> 0) & 0xFF).chr(($d >> 8) & 0xFF).chr($j)."\x0";
					break;
				}
			}
		}
        
		switch ($Locals_tmp { 0 }) {
		  
			case '!':
            
				$Locals_img = substr($Locals_tmp, 8, 10);
				$Locals_tmp = substr($Locals_tmp, 18, strlen($Locals_tmp) - 18);
                                
				break;
                
			case ',':
            
				$Locals_img = substr($Locals_tmp, 0, 10);
				$Locals_tmp = substr($Locals_tmp, 10, strlen($Locals_tmp) - 10);
                                
				break;
		}
        
		if (ord($this->frameSources[$i] { 10 }) & 0x80 && $this->imgBuilt) {
		  
			if ($Global_len == $Locals_len) {
			 
				if ($this->gifBlockCompare($Global_rgb, $Locals_rgb, $Global_len)) {
				    
					$this->gif .= $Locals_ext.$Locals_img.$Locals_tmp;
                    
				} else {
				    
					$byte = ord($Locals_img { 9 });
					$byte |= 0x80;
					$byte &= 0xF8;
					$byte |= (ord($this->frameSources[0] { 10 }) & 0x07);
					$Locals_img { 9 } = chr($byte);
					$this->gif .= $Locals_ext.$Locals_img.$Locals_rgb.$Locals_tmp;
				}
                
			} else {
			 
				$byte = ord($Locals_img { 9 });
				$byte |= 0x80;
				$byte &= 0xF8;
				$byte |= (ord($this->frameSources[$i] { 10 }) & 0x07);
				$Locals_img { 9 } = chr($byte);
				$this->gif .= $Locals_ext.$Locals_img.$Locals_rgb.$Locals_tmp;
			}
            
		} else {
		  
			$this->gif .= $Locals_ext.$Locals_img.$Locals_tmp;
		}
        
		$this->imgBuilt = true;
	}
    
	/**
	 * Add the gif string footer char
	 */
	protected function gifAddFooter()
	{
		$this->gif .= ';';
	}
    
	/**
	 * Compare two blocks and return 1 if they are equal, 0 if differ.
	 * 
	 * @param string $globalBlock
	 * @param string $localBlock
	 * @param integer $length
	 * 
	 * @return integer
	 */
	protected function gifBlockCompare($globalBlock, $localBlock, $length)
	{
		for ($i = 0; $i < $length; $i++) {
		  
			if ($globalBlock [ 3 * $i + 0 ] != $localBlock [ 3 * $i + 0 ] ||
			    $globalBlock [ 3 * $i + 1 ] != $localBlock [ 3 * $i + 1 ] ||
			    $globalBlock [ 3 * $i + 2 ] != $localBlock [ 3 * $i + 2 ]) {
				
				return 0;
			}
		}

		return 1;
	}
    
}

/**
 * Convert an integer to 2-byte little-endian binary data
 * 
 * @param integer $word Number to encode
 * 
 * @return string of 2 bytes representing @word as binary data
 */
function word2bin($word)
{
	return (chr($word & 0xFF).chr(($word >> 8) & 0xFF));
}
